packages/shared/src/vault/executeSwap.ts

```diff
@@ -1,58 +1,116 @@
+import { u8aToHex } from '@polkadot/util';
+import { decodeAddress } from '@polkadot/util-crypto';
 import { ContractTransactionReceipt } from 'ethers';
-import {
-  type ExecuteSwapParams,
-  executeSwapParamsSchema,
-  type NativeSwapParams,
-  type TokenSwapParams,
-  NativeCallParams,
-  TokenCallParams,
-} from './schemas';
 import { Vault__factory } from '../abis';
 import {
   checkAllowance,
   extractOverrides,
   getTokenContractAddress,
-  getVaultManagerContractAddress,
+  getVaultContractAddress,
   TransactionOptions,
 } from '../contracts';
-import { assetConstants, chainConstants, getInternalAsset } from '../enums';
-import { assert, isTokenCall, isTokenSwap } from '../guards';
-import { assertSignerIsConnectedToChain } from '../signer';
-import { SwapNetworkOptions } from './index';
+import {
+  assetConstants,
+  Chain,
+  chainConstants,
+  Chains,
+  getInternalAsset,
+  InternalAsset,
+} from '../enums';
+import { assertIsEvmChain, assertSignerIsConnectedToChain } from '../evm';
+import { assert } from '../guards';
+import { dotAddress } from '../parsers';
+import { ccmMetadataSchema } from '../schemas';
+import { Required } from '../types';
+import { assertValidAddress } from '../validation/addressValidation';
+import { ExecuteSwapParams, SwapNetworkOptions } from './index';
+
+const encodeAddress = (chain: Chain, address: string) => {
+  if (chain === Chains.Polkadot)
+    return u8aToHex(decodeAddress(dotAddress.parse(address)));
+  if (chain === Chains.Bitcoin)
+    return `0x${Buffer.from(address).toString('hex')}`;
+  if (chain === Chains.Ethereum) return address;
+
+  // no fallback encoding to prevent submitting txs with wrongly encoded addresses for new chains
+  throw new Error(`cannot encode address for chain ${chain}`);
+};
 
-const swapNative = async (
-  { destChain, destAsset, destAddress, amount }: NativeSwapParams,
-  networkOpts: SwapNetworkOptions,
-  txOpts: TransactionOptions,
-): Promise<ContractTransactionReceipt> => {
+export const assertIsValidSwap = (params: ExecuteSwapParams) => {
+  const internalSrcAsset = getInternalAsset({
+    chain: params.srcChain,
+    asset: params.srcAsset,
+  });
   const internalDestAsset = getInternalAsset({
-    chain: destChain,
-    asset: destAsset,
+    chain: params.destChain,
+    asset: params.destAsset,
   });
 
+  assert(
+    internalSrcAsset !== internalDestAsset,
+    `source asset and destination asset cannot be the same`,
+  );
+};
+
+const getVaultContract = (chain: Chain, networkOpts: SwapNetworkOptions) => {
   const vaultContractAddress =
     networkOpts.network === 'localnet'
       ? networkOpts.vaultContractAddress
-      : getVaultManagerContractAddress(networkOpts.network);
+      : getVaultContractAddress(chain, networkOpts.network);
+
```

Comment:
Should we also assert here that the address is defined?

Reply:
the `getVaultContractAddress` will throw if its not defined